gobblin-yarn/src/main/java/gobblin/yarn/GobblinApplicationMaster.java

```diff
@@ -0,0 +1,452 @@
+              HelixDataAccessor helixDataAccessor = helixManager.getHelixDataAccessor();
+
+              HelixProperty helixProperty = helixDataAccessor
+                  .getProperty(_message.getKey(helixDataAccessor.keyBuilder(), helixManager.getInstanceName()));
+              // The absence of the shutdown message indicates it has been removed
+              if (helixProperty == null) {
+                eventBus.post(new ApplicationMasterShutdownRequest());
+              }
+            }
+          }, 0, 1, TimeUnit.SECONDS);
+
+          result.setSuccess(true);
+          return result;
+        }
+
+        throw new RuntimeException(
+            String.format("Unknown %s message subtype: %s", Message.MessageType.SHUTDOWN.toString(), messageSubType));
+      }
+
+      @Override
+      public void onError(Exception e, ErrorCode code, ErrorType type) {
+        LOGGER.error(
+            String.format("Failed to handle message with exception %s, error code %s, error type %s", e, code, type));
+      }
+    }
+  }
+
+  /**
+   * A custom {@link MessageHandlerFactory} for {@link ControllerUserDefinedMessageHandler}s that
+   * handle messages of type {@link org.apache.helix.model.Message.MessageType#USER_DEFINE_MSG}.
+   */
+  private class ControllerUserDefinedMessageHandlerFactory implements MessageHandlerFactory {
+
+    @Override
+    public MessageHandler createHandler(Message message, NotificationContext context) {
+      return new ControllerUserDefinedMessageHandler(message, context);
+    }
+
+    @Override
+    public String getMessageType() {
+      return Message.MessageType.USER_DEFINE_MSG.toString();
+    }
+
+    @Override
+    public void reset() {
+
+    }
+
+    /**
+     * A custom {@link MessageHandler} for handling user-defined messages to the controller.
+     *
+     * <p>
+     *   Currently it handles the following sub types of messages:
+     *
+     *   <ul>
+     *     <li>{@link HelixMessageSubTypes#TOKEN_FILE_UPDATED}</li>
+     *   </ul>
+     * </p>
+     */
+    private class ControllerUserDefinedMessageHandler extends MessageHandler {
+
+      public ControllerUserDefinedMessageHandler(Message message, NotificationContext context) {
+        super(message, context);
+      }
+
+      @Override
+      public HelixTaskResult handleMessage() throws InterruptedException {
+        String messageSubType = this._message.getMsgSubType();
+
+        if (messageSubType.equalsIgnoreCase(HelixMessageSubTypes.TOKEN_FILE_UPDATED.toString())) {
+          LOGGER.info("Handling message " + HelixMessageSubTypes.TOKEN_FILE_UPDATED.toString());
+
+          eventBus.post(new DelegationTokenUpdatedEvent());
+          HelixTaskResult helixTaskResult = new HelixTaskResult();
+          helixTaskResult.setSuccess(true);
+          return helixTaskResult;
+        }
+
+        throw new RuntimeException(String.format("Unknown %s message subtype: %s",
```

Comment:
Seems more like an `IllegalArgumentException` or `IllegalStateException`. Since we provide the mapping between message type and message handler, Helix should also perform the mapping properly.